File: PageObjectDotNet/AutomationPractice/PageObjects/AutomationPracticeStore/Objects/QuickView.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.PageObjects;
using OpenQA.Selenium.Support.UI;
using SeleniumFramework.Base;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AutomationPractice.PageObjects.AutomationPracticeStore.Objects
{
    public class QuickView
    {

        [FindsBy(How = How.Id, Using = "bigpic")]
        IWebElement currentPicture;

        [FindsBy(How = How.CssSelector, Using = "ul#thumbs_list_frame li a")]
        IList<IWebElement> pictures;

        [FindsBy(How = How.Name, Using = "resetImages")]
        IWebElement displayAllPictures;

        [FindsBy(How = How.TagName, Using = "h1")]
        IWebElement productName;

        [FindsBy(How = How.CssSelector, Using = "p#product_reference span.editable")]
        IWebElement productModel;

        [FindsBy(How = How.CssSelector, Using = "p#product_condition span.editable")]
        IWebElement productCondition;

        [FindsBy(How = How.Id, Using = "short_description_content")]
        IWebElement productDescription;

        [FindsBy(How = How.Id, Using = "availability_date_value")]
        IWebElement productAvailability; //Available in the page but found in no product

        [FindsBy(How = How.Id, Using = "oosHook")]
        IWebElement productOutOfStock; //Not found something where this implementation is invoked

        [FindsBy(How = How.ClassName, Using = "socialsharing_product")]
        IWebElement socialButtonsHolder;

        [FindsBy(How = How.Id, Using = "our_price_display")]
        IWebElement labelPrice;

        [FindsBy(How = How.Id, Using = "reduction_percent_display")]
        IWebElement labelDiscount;

        [FindsBy(How = How.Id, Using = "old_price_display")]
        IWebElement labelOldPrice;

        [FindsBy(How = How.Id, Using = "quantity_wanted")]
        IWebElement inputQuantity;

        [FindsBy(How = How.ClassName, Using = "product_quantity_down")]
        IWebElement buttonQtySubtract;

        [FindsBy(How = How.ClassName, Using = "product_quantity_up")]
        IWebElement buttonQtyAdd;

        [FindsBy(How = How.Id, Using = "group_1")]
        IWebElement selectListSize;

        [FindsBy(How = How.Id, Using = "color_to_pick_list")]
        IWebElement listColor;

        [FindsBy(How = How.Name, Using = "Submit")]
        IWebElement buttonAddToCart;

        [FindsBy(How = How.Id, Using = "wishlist_button")]
        IWebElement buttonAddToWishlist;

        private readonly By popupMessage = By.ClassName("fancybox-skin");
        private readonly By frameContainer = By.CssSelector("iframe.fancybox-iframe");
        private By buttonCloseQuickView = By.ClassName("fancybox-close");

        private readonly IWebElement quickViewContainer;

        private readonly string ADD_TO_WISH_LIST_ERROR_NOT_LOGGED_IN = "You must be logged in to manage your wishlist.";


        public QuickView()
        {
            var driver = TestConfiguration.GetDriverInstance();
            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(8));
            wait.Until(ExpectedConditions.InvisibilityOfElementLocated(By.CssSelector("div#fancybox-loading")));
            wait.Until(ExpectedConditions.ElementExists(frameContainer));
            var iframe = driver.FindElement(frameContainer);
            driver.SwitchTo().Frame(iframe);
            quickViewContainer = driver.FindElement(By.Id("product"));
            PageFactory.InitElements(quickViewContainer, this);
        }

        public string GetCurrentPictureSource() { return currentPicture.GetAttribute("src"); }

        public IReadOnlyList<string> GetThumbnailPicturesSourceList()
        {
            return pictures.Where(img => img.Displayed).Select(img => img.GetAttribute("href")).ToList().AsReadOnly();
        }

        public string GetCurrentThumbnailPicture()
        {
            var found = pictures.Where(img => img.GetAttribute("class").Equals("fancybox shown")).FirstOrDefault()
                ?? throw new NoSuchElementException("There was a problem finding the currently selected thumbnail");
            return found.FindElement(By.TagName("img")).GetAttribute("src");
        }

        public QuickView DisplayAllThumbnailPictures()
        {
            try
            {
                displayAllPictures.Click();
                return this;
            }
            catch (NoSuchElementException ex)
            {
                throw new NoSuchElementException("Link not found or all available thumbnails may be already displayed", ex);
            }
        }

        public bool AreAllThumbnailPicturesDisplayed()
        {
            try
            {
                return !displayAllPictures.Displayed; }
            catch (NoSuchElementException)
            {
                return true;
            }
        }

        public string GetProductName() { return productName.Text; }

        public string GetProductModel() { return productModel.Text; }

        public string GetProductCondition() { return productCondition.Text; }

        public string GetProductDescription() { return productDescription.Text; }

        public float GetProductPrice() { return float.Parse(labelPrice.Text.Substring(1)); }

        public int GetProductDiscount() 
        {
            return string.IsNullOrEmpty(labelDiscount?.Text) ? 0 : int.Parse(labelDiscount.Text.Split('%')[0].Substring(1));
        }

        public float GetPriceBeforeDiscount() 
        {
            return string.IsNullOrEmpty(labelOldPrice?.Text) ? 0 : float.Parse(labelOldPrice.Text.Substring(1));
        }

        public int GetCurrentQuantity() { return int.Parse(inputQuantity.GetAttribute("value")); }

        public QuickView SetQuantity(string quantity) 
        {
            inputQuantity.Clear();
            inputQuantity.SendKeys(quantity);
            return this;
        }

        public IReadOnlyList<string> GetAvailableSizes() 
        {
            var options = new SelectElement(selectListSize).Options;
            return options.Select(opt => opt.Text).ToList().AsReadOnly();
        }

        public QuickView SetSize(string size) 
        {
            var selectElement = new SelectElement(selectListSize);
            selectElement.SelectByValue(size);
            return this;
        }

        public int IncrementQuantity() 
        {
            buttonQtyAdd.Click();
            return GetCurrentQuantity();
        }

        public int DecrementQuantity()
        {
            buttonQtySubtract.Click();
            return GetCurrentQuantity();
        }

        public IReadOnlyDictionary<string, string> GetColorsAvailable() 
        {
            return GetWebElementColorsAvailable().ToDictionary(colorWe => colorWe.GetAttribute("name"), colorWe => colorWe.GetAttribute("style").Split(':')[1].Trim());
        }

        public QuickView SelectColorByName(string color) 
        {
            SelectColor(color, true);
            return this;
        }

        public QuickView SelectColorByCode(string color)
        {
            SelectColor(color, false);
            return this;
        }

        public ProductInCartPopUp AddToCart() 
        {
            buttonAddToCart.Click();
            TestConfiguration.GetDriverInstance().SwitchTo().DefaultContent();
            return new ProductInCartPopUp();
        }

        public QuickView AddToWishList() 
        {
            buttonAddToWishlist.Click();

            try
            {
                var txt = GetPopupMessage();
                if (txt.Equals(ADD_TO_WISH_LIST_ERROR_NOT_LOGGED_IN))
                    throw new NotLoggedInException(txt);
            }
            catch (WebDriverTimeoutException ex)
            {
                throw new NoSuchElementException("The popup message after adding the item to the wish list did not appear", ex);
            }
            ClosePopupMessage();
            return this;
        }

        public string GetPopupMessage() { return GetPopupMessageElement().Text; }

        public QuickView ClosePopupMessage() 
        {
            var driver = TestConfiguration.GetDriverInstance();
            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(4));
            var closeBtn = wait.Until(ExpectedConditions.ElementToBeClickable(By.CssSelector("div.fancybox-skin a.fancybox-close")));
            closeBtn.Click();
            return this;
        }

        public void CloseQuickView() 
        {
            var driver = TestConfiguration.GetDriverInstance();
            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(3));
            driver.SwitchTo().DefaultContent();
            var closeButton = driver.FindElement(buttonCloseQuickView);
            closeButton.Click();
            wait.Until(ExpectedConditions.InvisibilityOfElementLocated(By.ClassName("fancybox-overlay")));
        }

        private IWebElement GetPopupMessageElement() 
        {
            var driver = TestConfiguration.GetDriverInstance();
            var wait = new WebDriverWait(driver, TimeSpan.FromSeconds(4));
            wait.Until(ExpectedConditions.ElementIsVisible(popupMessage));
            return driver.FindElement(popupMessage);
        }


        private IList<IWebElement> GetWebElementColorsAvailable()
        {
            return listColor.FindElements(By.TagName("a"));
        }

        private void SelectColor(string color, bool byName) 
        {
            var colorsAvailable = GetWebElementColorsAvailable();
            if (colorsAvailable.Count == 0)
                throw new NoSuchElementException("This QuickView window does not contain any color to choose from");

            var found = byName ? colorsAvailable.Where(we => string.Equals(we.GetAttribute("name"), color, StringComparison.CurrentCultureIgnoreCase)).FirstOrDefault() :
                colorsAvailable.Where(we => we.GetAttribute("style").ToLower().Contains(color.ToLower())).FirstOrDefault()
                ?? throw new NoSuchElementException($"The color '{color}' was not found among the available colors for product '{GetProductName()}'");
            found.Click();
        }

    }
}

File: PageObjectJava/src/test/java/PageObjects/AutomationPracticeStore/Objects/QuickView.java
package PageObjects.AutomationPracticeStore.Objects;

import Base.TestConfiguration;
import PageObjects.AutomationPracticeStore.NotLoggedInException;
import org.openqa.selenium.*;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.How;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory;
import org.openqa.selenium.support.pagefactory.ElementLocatorFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.*;
import java.util.function.Predicate;

public class QuickView  {

    @FindBy(id = "bigpic") WebElement currentPicture;
    @FindBy(css = "ul#thumbs_list_frame li a") List<WebElement> pictures;
    @FindBy(name = "resetImages") WebElement displayAllPictures;

    @FindBy(tagName = "h1") WebElement productName;
    @FindBy(css = "p#product_reference span.editable") WebElement productModel;
    @FindBy(css = "p#product_condition span.editable") WebElement productCondition;
    @FindBy(id = "short_description_content") WebElement productDescription;
    @FindBy(id = "availability_date_value") WebElement productAvailability; //Available in the page but found in no product
    @FindBy(id = "oosHook") WebElement productOutOfStock; //Not found something where this implementation is invoked
    @FindBy(className = "socialsharing_product") WebElement socialButtonsHolder;

    @FindBy(how = How.ID, using = "our_price_display") WebElement labelPrice;
    @FindBy(how = How.ID, using = "reduction_percent_display") WebElement labelDiscount;
    @FindBy(how = How.ID, using = "old_price_display") WebElement labelOldPrice;
    @FindBy(how = How.ID, using = "quantity_wanted") WebElement inputQuantity;
    @FindBy(how = How.CLASS_NAME, using = "product_quantity_down") WebElement buttonQtySubtract;
    @FindBy(how = How.CLASS_NAME, using = "product_quantity_up") WebElement buttonQtyAdd;
    @FindBy(how = How.ID, using = "group_1") WebElement selectListSize;
    @FindBy(how = How.ID, using = "color_to_pick_list") WebElement listColor;
    @FindBy(how = How.NAME, using = "Submit") WebElement buttonAddToCart;
    @FindBy(how = How.ID, using = "wishlist_button") WebElement buttonAddToWishlist;

    private final By popupMessage = By.className("fancybox-skin");
    private final By frameContainer = By.cssSelector("iframe.fancybox-iframe");
    private By buttonCloseQuickView = By.className("fancybox-close");

    private final WebElement container;

    private final String ADD_TO_WISH_LIST_ERROR_NOT_LOGGED_IN = "You must be logged in to manage your wishlist.";

    public QuickView(){
        WebDriverWait wait = new WebDriverWait(TestConfiguration.getDriver(), 8);
        wait.until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector("div#fancybox-loading")));
        wait.until(ExpectedConditions.presenceOfElementLocated(frameContainer));
        WebElement frame = TestConfiguration.getDriver().findElement(frameContainer);
        TestConfiguration.getDriver().switchTo().frame(frame);
        this.container = TestConfiguration.getDriver().findElement(By.id("product"));
        ElementLocatorFactory factory = new DefaultElementLocatorFactory(this.container);
        PageFactory.initElements(factory, this);
    }

    public String getCurrentPictureSource(){
        return this.currentPicture.getAttribute("src");
    }

    public List<String> getThumbnailPicturesSourceList(){
        List<String> thumbnailLinks = new ArrayList<>();
        this.pictures.stream().filter(WebElement::isDisplayed).forEach(pic -> thumbnailLinks.add(pic.getAttribute("href")));
        return thumbnailLinks;
    }

    public String getCurrentThumbnailPicture(){
        WebElement selectedThumbnail = this.pictures.stream().filter(pic -> pic.getAttribute("class").equalsIgnoreCase("fancybox shown"))
                .findFirst().orElseThrow(() -> new NoSuchElementException("There was a problem finding the currently selected thumbnail"));
        return selectedThumbnail.findElement(By.tagName("img")).getAttribute("src");
    }

    public QuickView displayAllThumbnailPictures(){
        try{
            displayAllPictures.click();
            return this;
        }catch(NoSuchElementException ex){
            throw new NoSuchElementException("All available thumbnails may be already displayed", ex.fillInStackTrace());
        }
    }

    public boolean areAllThumbnailPicturesDisplayed(){
        try{
            return !displayAllPictures.isDisplayed();
        }catch(NoSuchElementException ex){
            return true;
        }
    }

    public String getProductName(){
        return productName.getText();
    }

    public String getProductModel(){
        return productModel.getText();
    }

    public String getProductCondition(){
        return productCondition.getText();
    }

    public String getProductDescription(){
        return productDescription.getText();
    }

    public double getProductPrice(){
        return Double.parseDouble(labelPrice.getText().substring(1));
    }

    public int getProductDiscount(){
        if (Objects.isNull(labelDiscount) || labelDiscount.getText().isEmpty())
            return 0;
        return Integer.parseInt(labelDiscount.getText().split("%")[0].substring(1));
    }

    public double getPriceBeforeDiscount(){
        if (Objects.isNull(labelOldPrice) || labelOldPrice.getText().isEmpty())
            return 0.0;
        return Double.parseDouble(labelOldPrice.getText().substring(1));
    }

    public int getCurrentQuantity(){
        return Integer.parseUnsignedInt(inputQuantity.getAttribute("value"));
    }

    public QuickView setQuantity(String quantity){
        inputQuantity.clear();
        inputQuantity.sendKeys(quantity);
        return this;
    }

    public List<String> getAvailableSizes(){
        List<String> sizes = new ArrayList<>();
        List<WebElement> options = new Select(this.selectListSize).getOptions();
        options.forEach(itm -> sizes.add(itm.getText()));
        return sizes;
    }

    public QuickView setSize(String size){
        Select options = new Select(this.selectListSize);
        options.selectByValue(size);
        return this;
    }

    public int incrementQuantity(){
        buttonQtyAdd.click();
        return this.getCurrentQuantity();
    }

    public int decrementQuantity(){
        buttonQtySubtract.click();
        return this.getCurrentQuantity();
    }

    public Map<String,String> getColorsAvailable(){
        Map<String,String> colors_map = new HashMap<>();
        this.getWebElementColorsAvailable().forEach(color -> {
            colors_map.put(color.getAttribute("name"), color.getAttribute("style").split(": ")[1]);
        });
        return colors_map;
    }

    public QuickView selectColorByName(String color){
        this.selectColor(color, true);
        return this;
    }

    public QuickView selectColorByCode(String color){
        this.selectColor(color, false);
        return this;
    }

    public ProductInCartPopUp addToCart(){
        this.buttonAddToCart.click();
        TestConfiguration.getDriver().switchTo().defaultContent();
        return new ProductInCartPopUp();
    }

    public QuickView addToWishList(){
        this.buttonAddToWishlist.click();
        try{
            String text = this.getPopupMessage();
            if(text.equals(ADD_TO_WISH_LIST_ERROR_NOT_LOGGED_IN))
                throw new NotLoggedInException(text);
        }catch(TimeoutException ex){
            throw new NoSuchElementException("The pop up message after adding the item to the wish list did not appear", ex.fillInStackTrace());
        }
        this.closePopupMessage();
        return this;
    }

    public String getPopupMessage() throws TimeoutException{
        WebDriverWait wait = new WebDriverWait(TestConfiguration.getDriver(), 4);
        wait.until(ExpectedConditions.visibilityOfElementLocated(popupMessage));
        return TestConfiguration.getDriver().findElement(popupMessage).getText();
    }

    public QuickView closePopupMessage(){
        WebDriverWait wait = new WebDriverWait(TestConfiguration.getDriver(), 4);
        WebElement closeButton = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("div.fancybox-skin a.fancybox-close")));
        closeButton.click();
        return this;
    }

    public void closeQuickView(){
        TestConfiguration.getDriver().switchTo().defaultContent();
        WebDriverWait wait = new WebDriverWait(TestConfiguration.getDriver(), 3);
        WebElement button = TestConfiguration.getDriver().findElement(buttonCloseQuickView);
        button.click();
        wait.until(ExpectedConditions.invisibilityOfElementLocated(By.className("fancybox-overlay")));
    }


    private List<WebElement> getWebElementColorsAvailable(){
        return listColor.findElements(By.tagName("a"));
    }

    private void selectColor(String color, boolean byName){
        if (this.getWebElementColorsAvailable().size() == 0)
            throw new NoSuchElementException("This QuickView window does not contain any color to choose from");

        Predicate<WebElement> findByName = itm -> itm.getAttribute("name").equalsIgnoreCase(color);
        Predicate<WebElement> findByCode = itm -> this.containsIgnoreCase(itm.getAttribute("style"), color);

        WebElement found;
        if(byName)
            found = this.getWebElementColorsAvailable().stream().filter(findByName).findFirst().orElse(null);
        else
            found = this.getWebElementColorsAvailable().stream().filter(findByCode).findFirst().orElse(null);

        if(Objects.isNull(found))
            throw new NoSuchElementException(String.format("The color '%s' was not found among the available colors for product %s", color, this.getProductName()));
        else
            found.click();
    }

    private boolean containsIgnoreCase(String str, String subString) {
        return str.toLowerCase().contains(subString.toLowerCase());
    }
}
